File: runme/mcgen/local/datan/solsrc/s2min.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern double CBFUNC minbwl(double const P_T *x, integer const P_T *n);
extern void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                         double P_T *r, integer const P_T *n);

/* Common Block Declarations */

struct cminbw {
    double y[1000];
    integer nny;
} LIBDATA cminbw;

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__74 = 74,
               c__1000 = 1000;
static double c_dp15 = .15, c_dp85 = .85, c_dmp414 = -.414, c_d1 = 1.,
              c_d0 = 0., c_dmp2 = -.2, c_dp9 = .9;

main(void)
{
    /* Local variables */
    static integer i, k, ipl, list[2], nseed1, nseed2, nstep, nws;
    static double cx[4] /* was [2][2] */, dpl, epsiln, fact, dels, fmin,
                  gamma, x[2], delxm, scrat[6] /* was [2][3] */, s2, a,
                  scrat1[4] /* was [2][2] */, xmin, xmax, ymin, ymax, dels2,
                  xpl[1000], ypl[1000];
    static logical ok;
    static char string[75];
    /* System generated locals */
    integer i__1;
    double d__1, d__2, d__3;
/* identify program to user */
    printf("%s\n", "Program S2MIN performs fit of Breit-Wigner");
    printf("%s\n\n", "to small sample by minimization.");
/* ask for input */
    printf("%s\n", "Enter number n of events (2 <= n <= 1000):");
    printf("%s", "> ");
    scanf("%li", &cminbw.nny);
/* draw sample from Breit-Wigner distribution */
    nseed1 = 87654;
    nseed2 = 98765;
    rne2in_(&nseed1, &nseed2);
    a = 0.;
    gamma = 1.;
    rnbw(&a, &gamma, cminbw.y, &cminbw.nny);
/* write table of data */
    printf("%s\n", "sample Y is");
    for (k = 1; k <= cminbw.nny; ++k) {
        if (k % 10 == 1)
            printf(" ");
        printf("%7.2lf", cminbw.y[k-1]);
        if (k % 10 == 0)
            printf("\n");
    }
    if (cminbw.nny % 10 != 0)
        printf("\n");
/* find first approximation */
    smmnvr_(cminbw.y, &cminbw.nny, x, &delxm, &s2, &dels2, &x[1], &dels);
/* minimize with MINSIM */
    list[0] = 1;
    list[1] = 1;
    nstep = 10000;
    epsiln = 0.;
    fmin = .01;
    printf("%s\n", "minimization with MINSIM");
    printf("%s%4li%s%4li%s%4li%s%2li%2li\n", "N = ", cminbw.nny,
           ", NR = ", c__2, ", NRED = ", c__2, ", LIST = ", list[0], list[1]);
    printf("%s%10.5lf%10.5lf\n", "first approx.: X = ", x[0], x[1]);
    minsim_(x, &c__2, &c__2, list, minbwl, &fmin, &epsiln, &nstep, scrat);
    printf("%s%6.2lf%s%6li\n%s%10.5lf%10.5lf\n",
           "result of minimization: FMIN = ", fmin, ", NSTEP =", nstep,
           "X =", x[0], x[1]);
/* determine covariance matrix */
    fact = 1.;
    mincov_(x, &c__2, &c__2, list, minbwl, &fact, scrat, scrat1, cx, &ok);
    if (! ok) {
        printf("%s\n", "determination of covariance matrix fails");
        exit(0);
    }
    printf("%s\n", "covariance matrix CX = ");
    mtxwrt_(cx, &c__2, &c__2);
    printf("\n");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* prepare graphics */
    xmin = -10.;
    xmax = 10.;
    ymin = 0.;
    ymax = 1.;
    gropen_();
    gropws_(&nws);
    grwncc_(&xmin, &xmax, &ymin, &ymax);
    grvwwc_(&c_dmp2, &c_dp9, &c_dp15, &c_dp85);
    grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
    grfram_();
    grboun_();
    sprintf(string, "N = %4li, x_1# = %6.3lf, x_2# = %6.3lf"
            ", &D@x_1# = %6.3lf, &D@x_2# = %6.3lf", cminbw.nny, x[0], x[1],
            sqrt(cx[0]), sqrt(cx[3]));
    string[74] = ' ';
    grtxtc_(&c_d1, string, &c__74);
    memset(string, ' ', 75);
    strncpy(string, "x_1", 3);
    grsclx_(string, &c__3);
    strncpy(string, "x_2", 3);
    grscly_(string, &c__3);
/* plot scatter diagram */
    grstcl_(&c__2);
    i__1 = cminbw.nny;
    for (i = 1; i <= i__1; ++i) {
        xpl[0] = cminbw.y[i - 1];
        xpl[1] = xpl[0];
        ypl[0] = 0.;
        ypl[1] = .1;
        grplin_(&c__2, xpl, ypl);
    }
/* draw Breit-Wigner corresponding to solution */
    fact = 2. / (x[1] * 3.14159);
    dpl = (xmax - xmin) / 999.;
    for (ipl = 1; ipl <= 1000; ++ipl) {
        xpl[ipl - 1] = xmin + (double) (ipl - 1) * dpl;
/* Computing 2nd power */
        d__1 = x[1];
/* Computing 2nd power */
        d__2 = xpl[ipl - 1] - x[0];
/* Computing 2nd power */
        d__3 = x[1];
        ypl[ipl - 1] = fact * (d__1 * d__1) / (d__2 * d__2 * 4. + d__3 * d__3);
    }
    grstcl_(&c__4);
    grplin_(&c__1000, xpl, ypl);
    grclse_();
    return 0;
} /* main */

/* -------------------------------------------- */
double CBFUNC minbwl(double const P_T *x, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;
    double ret_val, d__1, d__2;

    /* Local variables */
    static double f;
    static integer i;

    /* Parameter adjustments */
    --x;

    /* Function Body */
    f = (double) cminbw.nny * (log(2.) - log(3.14159) + log((abs(x[2]))))
            ;
    i__1 = cminbw.nny;
    for (i = 1; i <= i__1; ++i) {
/* Computing 2nd power */
        d__1 = cminbw.y[i - 1] - x[1];
/* Computing 2nd power */
        d__2 = x[2];
        f -= log(d__1 * d__1 * 4. + d__2 * d__2);
    }
    ret_val = -f;
    return ret_val;
} /* minbwl */

/* ------------------------------------------------------------------ */
void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                  double P_T *r, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */
    static integer i;

    /* Parameter adjustments */
    --r;

    /* Function Body */
    rnecuy_(&r[1], n);
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        r[i] = *a + *gamma * .5 * tan((r[i] - .5) * 3.14159);
    }
    return;
} /* rnbw */

File: runme/mcgen/local/datan/solsrc/s2min.f
      PROGRAM S2MIN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /CMINBW/ Y,NNY
      DIMENSION X(2),Y(1000),XPL(1000),YPL(1000),LIST(2)
      DIMENSION CX(2,2),SCRAT(2,3),SCRAT1(2,2)
      PARAMETER(NR=2,NRED=2,NPL=1000,PI=3.14159D0)
      DOUBLE PRECISION MINBWL
      CHARACTER*75 STRING
      LOGICAL OK
      EXTERNAL MINBWL
C identify program to user
      WRITE(*,*)' program S2MIN performs fit of Breit-Wigner '
      WRITE(*,*)' to small sample by minimization'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number n of events(2 <= n <= 1001)'
      WRITE(*,*)' >'
      READ(*,*)NNY
C draw sample from Breit-Wigner distribution
      NSEED1=87654
      NSEED2=98765
      CALL RNE2IN(NSEED1,NSEED2)
      A=0.D0
      GAMMA=1.D0
      CALL RNBW(A,GAMMA,Y,NNY)
C write table of data
      WRITE(*,*)' sample Y is'
      WRITE(*,'(1X,10F7.2)')(Y(K),K=1,NNY)
      WRITE(*,'(/)')
C find first approximation
      CALL SMMNVR(Y,NNY,X(1),DELXM,S2,DELS2,X(2),DELS)
C minimize with MINSIM
      LIST(1)=1
      LIST(2)=1
      NSTEP=10000
      EPSILN=0.D0
      FMIN=.01D0
      WRITE(*,*)' minimization with MINSIM'
      WRITE(*,'(3(A,I4),A,2I2)')' N = ',NNY,', NR = ',NR,', NRED = ' ,
     +NRED,', LIST = ',LIST
      WRITE(*,'(A,2F10.5)')' first approx.: X = ',X
      CALL MINSIM(X,NR,NRED,LIST,MINBWL,FMIN,EPSILN,NSTEP,SCRAT)
      WRITE(*,'(A,F6.2,A,I6,/,A,(2F10.5))')
     +' result of minimization: FMIN = ', FMIN,', NSTEP =',NSTEP,' X =',
     +X
C determine covariance matrix
      FACT=DBLE(1.)
      CALL MINCOV(X,NR,NRED,LIST,MINBWL,FACT,SCRAT,SCRAT1,CX,OK)
      IF(.NOT.OK) THEN
        WRITE(*,*)' determination of covariance matrix fails'
        STOP
      END IF
      WRITE(*,*)' covariance matrix CX = '
      CALL MTXWRT(CX,NRED,NRED)
      WRITE(*,'(/)')
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C prepare graphics
      XMIN=-10.D0
      XMAX=10.D0
      YMIN=0.D0
      YMAX=1.D0
      CALL GROPEN
      CALL GROPWS(NWS)
      CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
      CALL GRVWWC(-0.2D0,0.9D0,0.15D0,0.85D0)
      CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
      CALL GRFRAM
      CALL GRBOUN
      STRING(1:38)='N = ****, x_1# = ******, x_2# = ******'
      STRING(39:74)=', &D@x_1# = ******, &D@x_2# = ******'
      WRITE(STRING(5:8),'(I4)') NNY
      WRITE(STRING(18:23),'(F6.3)') X(1)
      WRITE(STRING(33:38),'(F6.3)') X(2)
      WRITE(STRING(51:56),'(F6.3)') SQRT(CX(1,1))
      WRITE(STRING(69:74),'(F6.3)') SQRT(CX(2,2))
      CALL GRTXTC(1.D0,STRING,74)
      STRING='x_1'
      CALL GRSCLX(STRING,3)
      STRING='x_2'
      CALL GRSCLY(STRING,3)
C plot scatter diagram
      CALL GRSTCL(2)
      DO 20 I=1,NNY
        XPL(1)=Y(I)
        XPL(2)=XPL(1)
        YPL(1)=0.D0
        YPL(2)=0.1D0
        CALL GRPLIN(2,XPL,YPL)
20    CONTINUE
C draw Breit-Wigner corresponding to solution
      FACT=2.D0/(PI*X(2))
      DPL=(XMAX-XMIN)/DBLE(NPL-1)
      DO 30 IPL=1,NPL
        XPL(IPL)=XMIN+DBLE(IPL-1)*DPL
        YPL(IPL)=FACT*X(2)**2/(4.D0*(XPL(IPL)-X(1))**2+X(2)**2)
30    CONTINUE
      CALL GRSTCL(4)
      CALL GRPLIN(NPL,XPL,YPL)
      CALL GRCLSE
      END
C--------------------------------------------
      DOUBLE PRECISION FUNCTION MINBWL(X,N)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      COMMON /CMINBW/ Y,NY
      DIMENSION Y(1000)
      DIMENSION X(N)
      PARAMETER(PI=3.14159D0)
      F=DBLE(NY)*(LOG(2.D0)-LOG(PI)+LOG(ABS(X(2))))
      DO 10 I=1,NY
        F=F-LOG(4.D0*(Y(I)-X(1))**2+X(2)**2)
10    CONTINUE
      MINBWL=-F
      END
C ------------------------------------------------------------------
      SUBROUTINE RNBW(A,GAMMA,R,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(PI=3.14159D0,HALF=0.5D0)
      DIMENSION R(N)
      CALL RNECUY(R,N)
      DO 10 I=1,N
        R(I)=A+HALF*GAMMA*TAN(PI*(R(I)-HALF))
10    CONTINUE
      END
